client/mm/price_oracle.go

```diff
@@ -39,38 +48,110 @@ type cachedPrice struct {
 	mtx     sync.RWMutex
 	stamp   time.Time
 	price   float64
-	oracles []*oracleReport
+	oracles []*OracleReport
 
 	base  *asset.RegisteredAsset
 	quote *asset.RegisteredAsset
 }
 
 // priceOracle periodically fetches market prices from a set of oracles.
 type priceOracle struct {
-	ctx          context.Context
-	log          dex.Logger
-	cachedPrices map[string]*cachedPrice
+	ctx      context.Context
+	log      dex.Logger
+	autoSync bool
+
+	cachedPricesMtx sync.RWMutex
+	cachedPrices    map[string]*cachedPrice
 }
 
 type oracle interface {
-	getMarketPrice(base, quote uint32) float64
+	GetMarketPrice(base, quote uint32) float64
 }
 
 var _ oracle = (*priceOracle)(nil)
 
-func (o *priceOracle) getMarketPrice(base, quote uint32) float64 {
+func (o *priceOracle) getOracleDataAutoSync(base, quote uint32) (float64, []*OracleReport, error) {
 	mktStr := (&mkt{base, quote}).String()
 
-	if price, ok := o.cachedPrices[mktStr]; ok {
+	o.cachedPricesMtx.RLock()
+	price, ok := o.cachedPrices[mktStr]
+	o.cachedPricesMtx.RUnlock()
+
+	if ok {
 		price.mtx.RLock()
 		defer price.mtx.RUnlock()
 
-		if time.Since(price.stamp) < oraclePriceExpiration {
-			return price.price
+		if o.autoSync && time.Since(price.stamp) < oraclePriceExpiration {
+			return price.price, price.oracles, nil
+		}
+	}
+
+	return 0, nil, fmt.Errorf("expired price data for %s", mktStr)
+}
+
+func (o *priceOracle) getOracleDataNoAutoSync(base, quote uint32) (float64, []*OracleReport, error) {
+	mktStr := (&mkt{base, quote}).String()
+
+	o.cachedPricesMtx.Lock()
+	price, ok := o.cachedPrices[mktStr]
+	if !ok {
+		cp, err := newCachedPrice(base, quote, asset.Assets())
+		if err != nil {
+			o.cachedPricesMtx.Unlock()
+			return 0, nil, err
 		}
+		o.cachedPrices[mktStr] = cp
+		price = cp
+	}
+	o.cachedPricesMtx.Unlock()
+
+	price.mtx.RLock()
+	if time.Since(price.stamp) < oracleRecheckInterval {
+		o.log.Infof("priceOracle: within recheck interval %v", mktStr)
+		price.mtx.RUnlock()
+		return price.price, price.oracles, nil
```

Comment:
Is only price.stamp under the lock? These are read after the lock is unlocked I believe.